networkx/generators/random_graphs.py

```diff
@@ -962,3 +963,79 @@ def random_powerlaw_tree_sequence(n, gamma=3, seed=None, tries=100):
           "Exceeded max (%d) attempts for a valid tree sequence."%tries)
     return False
 
+
+def random_kernel_graph(n, kernel_integral, kernel_root=None, seed=None):
+    """Return an random graph based on the specified kernel.
+
+    The algorithm chooses each of the `[n(n-1)]/2` possible edges with
+    probability specified by a kernel `\kappa(x,y)` [1]_.  The kernel 
+    `\kappa(x,y)` must be a symmetric (in x,y), non-negative, bounded function.
+
+    Parameters
+    ----------
+    n : int
+        The number of nodes
+    kernal_integral : function
+        Function that returns the definite integral of the kernel `\kappa(x,y)`,
+        `F(y,a,b) := \int_a^b \kappa(x,y)dx`
+    kernel_root: function (optional)
+        Function that returns the root `b` of the equation `F(y,a,b) = r`.
+        If None, the root is found using scipy.optimize.brentq (requires scipy).
+    seed : int, optional
+        Seed for random number generator (default=None)
+
+    Notes
+    -----
+    The kernel is specified through its definite integral which must be 
+    provided as one of the arguments. If the integral and root of the 
+    kernal integral can be found in O(1) time then this algorithm runs in 
+    time O(n+m) where m is the expected number of edges [2]_.
+
+    The nodes are set to integers from 0 to n-1.
+
+    Examples
+    --------
+    Generate an Erdős–Rényi random graph `G(n,c/n)`, with kernel
+    `\kappa(x,y)=c` where `c` is the mean expected degree.
+
+    >>> def integral(u, w, z):
+    ...     return c*(z-w)
+    >>> def root(u, w, r):
+    ...     return r/c+w
+    >>> c = 1.0
+    >>> graph = random_kernel_graph(1000, integral, root)
+
+    See Also
+    --------
+    gnp_random_graph
+    expected_degree_graph
+
+    References
+    ----------
+    .. [1] Bollobás, Béla,  Janson, S. and Riordan, O.
+       "The phase transition in inhomogeneous random graphs",
+       Random Structures Algorithms, 31, 3-122, 2007.
+
+    .. [2] Aric Hagberg and Nathan Lemons
+       "Fast generation of sparse random kernel graphs"
+       Unpublished, in review.
+    """
+    if not seed is None:
+        random.seed(seed)
+    if kernel_root is None:
+        import scipy.optimize as optimize
+        def kernel_root(y, a, r):
+            def my_function(b):
+                return kernel_integral(y, a, b) - r
+            return optimize.brentq(my_function, a, 1)
+    graph = nx.Graph()
+    graph.add_nodes_from(range(n))
+    (i, j) = (1, 1)
+    while i < n:
+        r = -math.log(1 - random.random()) # (1-random.random()) in (0, 1]
+        if kernel_integral(float(i)/n, float(j)/n, 1.0) <= r:
```

Comment:
Yes, imports from `__future__` must come first.

If it breaks other functions, they were already broken on Python 2 by relying on `/` computing integer division if its operands are integers instead of using real floating-point division.